### PR TITLE
testcase/filesystem : add scenario tc for filesystem

### PR DESCRIPTION
**Pull request by ktejram**
file path : apps/examples/testcase/le_tc/filesystem
file name : fs_main.c

scenario tc :
tc_fs_vfs_umount_twice_n
tc_fs_vfs_mount_read_mode_p
tc_fs_vfs_mount_twice_n
tc_fs_vfs_open_n
tc_fs_vfs_close_n
tc_fs_vfs_read_empty_file_p
tc_fs_vfs_write_read_p
tc_fs_vfs_fsync_p
tc_fs_vfs_lseek_n
tc_fs_vfs_lseek_closed_fd_n
tc_fs_vfs_lseek_p
tc_fs_vfs_pwrite_pread_p
tc_fs_vfs_mkdir_n
tc_fs_vfs_rmdir_n
tc_fs_vfs_mkdir_rmdir_n
tc_fs_vfs_opendir_invalid_path_n
tc_fs_vfs_opendir_file_path_n
tc_fs_vfs_readdir_rewinddir_p
tc_fs_vfs_seekdir_telldir_p
tc_fs_vfs_closedir_n
tc_libc_stdio_fdopen_p
tc_libc_stdio_fdopen_n
tc_libc_stdio_fopen_invalid_path_n
tc_libc_stdio_fopen_invalid_mode_n
tc_libc_stdio_fclose_null_n
tc_libc_stdio_fputs_n
tc_libc_stdio_fgets_n
tc_libc_stdio_fputs_fgets_p
tc_libc_stdio_fseek_n
tc_libc_stdio_fseek_invalid_fp_n
tc_libc_stdio_ftell_n
tc_libc_stdio_fseek_ftell_p
tc_libc_stdio_fwrite_fread_p

Signed-off-by: gaurav.m2 <gaurav.m2@samsung.com>

### CODE REVIEW

**Comment by seinfra** [comment]
Code Rule Check Result:
apps/examples/testcase/le_tc/filesystem/fs_main.c:3045: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3088: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/le_tc/filesystem/fs_main.c:3095: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3109: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/le_tc/filesystem/fs_main.c:3177: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/le_tc/filesystem/fs_main.c:3219: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/le_tc/filesystem/fs_main.c:3266: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3269: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3273: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3274: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3357: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/le_tc/filesystem/fs_main.c:3436: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/le_tc/filesystem/fs_main.c:3501: ERROR: [SPC_M_SEP] space prohibited before that close parenthesis ')'
apps/examples/testcase/le_tc/filesystem/fs_main.c:3502: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3664: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3669: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3678: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3681: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3684: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3687: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3692: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3698: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3737: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3741: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3744: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3747: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3748: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3751: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3755: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3758: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3761: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3762: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3765: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3769: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3772: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3775: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3778: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3782: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3785: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3788: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3789: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3792: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3906: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)

**Comment by seinfra** [comment]
Code Rule Check Result:
apps/examples/testcase/le_tc/filesystem/fs_main.c:3045: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3095: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3266: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3269: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3273: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3274: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3502: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3664: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3669: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3678: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3681: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3684: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3687: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3692: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3698: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3737: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3741: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3744: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3747: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3748: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3751: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3755: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3758: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3761: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3762: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3765: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3769: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3772: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3775: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3778: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3782: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3785: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3788: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3789: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3792: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)
apps/examples/testcase/le_tc/filesystem/fs_main.c:3906: ERROR: [SPC_M_OPR] space required after that ';' (ctx:VxV)

**Comment by Taejun-Kwon** [comment]
Please do not apply scenario tc, this tc is only for checking api.

**Comment by seinfra** [comment]
Code Rule Check OK.

File: apps/examples/testcase/le_tc/filesystem/itc_fs.c [hunk 1]
```diff
@@ -0,0 +1,1442 @@
...
+*/
+static void add_dirs(int count)
+{
+	char folder_name[BUFFER_LEN];
+	int len;
+	int i;
+	strcpy(folder_name, VFS_FOLDER_PATH);
+	strcat(folder_name, "/0\0");
+	len = strlen(folder_name);
+	mkdir(VFS_FOLDER_PATH, 0777);
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/le_tc/filesystem/itc_fs.c]: here it makes parent folder, then return should be checked.

File: apps/examples/testcase/le_tc/filesystem/itc_fs.c [hunk 2]
```diff
@@ -0,0 +1,1442 @@
...
+	for (i = 0; i < count; i++) {
+		folder_name[len - 1] = '0' + i;
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/le_tc/filesystem/itc_fs.c]: How about make function like 'generate_childdir(char* dir, int num)' add_dirs & remove_dirs can use it.

Comment (Taejun-Kwon) [reply apps/examples/testcase/le_tc/filesystem/itc_fs.c]: also there's itoa in TizenRT

File: apps/examples/testcase/le_tc/filesystem/itc_fs.c [hunk 3]
```diff
@@ -0,0 +1,1442 @@
...
+*/
+static void itc_fs_vfs_umount_n_twice(void)
+{
+	int ret;
+
+	ret = umount(CONFIG_MOUNT_POINT);
+	TC_ASSERT_EQ("umount", ret, OK);
+
+	ret = umount(CONFIG_MOUNT_POINT);
+	TC_ASSERT_EQ("umount", ret, ERROR);
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/le_tc/filesystem/itc_fs.c]: NEQ OK is more general usage.

File: apps/examples/testcase/le_tc/filesystem/itc_fs.c [hunk 4]
```diff
@@ -0,0 +1,1442 @@
...
+* @scenario         Mount file system in read mode, and try to create directory.
+* @apicovered       mount
+* @precondition     File system should be initialized and not mounted.
+* @postcondition    NA
+*/
+static void itc_fs_vfs_mount_p_read_mode(void)
+{
+	int ret;
+
+	ret = mount(MOUNT_DEV_DIR, CONFIG_MOUNT_POINT, TARGET_FS_NAME, MS_RDONLY, NULL);
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/le_tc/filesystem/itc_fs.c]: mount function use FAR const char *target as a second parameter, you'd better rename TARGET_FS_NAME to FS_TYPE

File: apps/examples/testcase/le_tc/filesystem/itc_fs.c [hunk 5]
```diff
@@ -0,0 +1,1442 @@
...
+* @brief            Syncs the file in buffer with the disk.
+* @scenario         Open same file with two file descriptors in read and write mode, and checks fsync.
+* @apicovered       fsync, write, read
+* @precondition     NA
+* @postcondition    NA
+*/
+static void itc_fs_vfs_fsync_p(void)
+{
+	int fd;
+	int fd1;
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/le_tc/filesystem/itc_fs.c]: fd1, fd2

File: apps/examples/testcase/le_tc/filesystem/itc_fs.c [hunk 6]
```diff
@@ -0,0 +1,1442 @@
...
+	int ret;
+	int len;
+	char *write_buf = VFS_TEST_CONTENTS_1;
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/le_tc/filesystem/itc_fs.c]: you write once, you don't have to use 2 buffers.
char buf[BUFFER_LEN];

int len = strlen(VFS_TEST_CONTENTS_1);
write(fd, buf, len);
.
.
.
memset(buf)
and then read.

Comment (Taejun-Kwon) [reply apps/examples/testcase/le_tc/filesystem/itc_fs.c]: please check other place too.

File: apps/examples/testcase/le_tc/filesystem/itc_fs.c [hunk 7]
```diff
@@ -0,0 +1,1442 @@
...
+* @brief            Move current file position to specific position
+* @scenario         Move file position with invalid file descriptor(fd), invalid whence.
+* @apicovered       lseek
+* @precondition     NA
+* @postcondition    NA
+*/
+static void itc_fs_vfs_lseek_n(void)
+{
+	int fd;
+	int ret;
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/le_tc/filesystem/itc_fs.c]: lseek's type is off_t

File: apps/examples/testcase/le_tc/filesystem/itc_fs.c [hunk 8]
```diff
@@ -0,0 +1,1442 @@
...
+
+	fd = open(VFS_FILE_PATH, O_RDWR | O_TRUNC | O_CREAT);
+	TC_ASSERT_GEQ("open", fd, 0);
+	ret = close(fd);
+	TC_ASSERT_EQ("close", ret, OK);
+
+	ret = lseek(fd, 0, SEEK_SET);
+	TC_ASSERT_EQ("lseek", ret, ERROR);
```
Comment (Taejun-Kwon) [inline apps/examples/testcase/le_tc/filesystem/itc_fs.c]: TC_ASSERT_EQ("lseek" lseek(fd, 0, SEEK_SET), (off_t)ERROR);

Comment (Taejun-Kwon) [reply apps/examples/testcase/le_tc/filesystem/itc_fs.c]: There are many place where you can reduce code line.

**Comment by ktejram** [comment]
Resolved merge conflicts and fix review comments in new commit. Also done ITCs name change.
new commit PR : https://github.com/Samsung/TizenRT/pull/1082

**Pull request closed without merging**